### ISSUE TITLE
Synchronized access to MessageService restricts performance

### ISSUE THREAD

**Issue by magnusbaeck** [eiffel-community/eiffel-remrem-publish#296]
### Description

The [ProducerController](https://github.com/eiffel-community/eiffel-remrem-publish/blob/fe2e46783444fe4674a0a01a8a585c738f31a0ec/publish-service/src/main/java/com/ericsson/eiffel/remrem/publish/controller/ProducerController.java) synchronizes all calls to the [MessageService](https://github.com/eiffel-community/eiffel-remrem-publish/blob/fe2e46783444fe4674a0a01a8a585c738f31a0ec/publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/service/MessageService.java). This serializes all publish operations and thus has severe performance implications when the rate of incoming requests reaches a certain point. In our case our requests usually take about 35 ms but during our peak hours publishing a single message can take several seconds. This increases the total execution time for our ETOS executions significantly.

### Additional Context

The problematic synchronization was added in #219 in an attempt to address a problem with corrupt HTTP responses without addressing the root cause of the poor implementation of [MessageServiceRMQImpl](https://github.com/eiffel-community/eiffel-remrem-publish/blob/fe2e46783444fe4674a0a01a8a585c738f31a0ec/publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/service/MessageServiceRMQImpl.java). My code review comment in said PR (https://github.com/eiffel-community/eiffel-remrem-publish/pull/219#discussion_r743612694) explains the problem but was never addressed.

The following graphs illustrate what we're seeing every night. Note how the request latency shoots up by several orders of magnitude without being constrained by thread pools, CPU, or memory. (The CPU usage on the right y axis of the top-left graph is in millicores.)

![image](https://github.com/user-attachments/assets/baf56e84-f74d-4ab1-bfc2-3eceec35a5ad)

This serialization also effectively disables the attempt at a connection pool in [RabbitMqProperties](https://github.com/eiffel-community/eiffel-remrem-publish/blob/fe2e46783444fe4674a0a01a8a585c738f31a0ec/publish-common/src/main/java/com/ericsson/eiffel/remrem/publish/helper/RabbitMqProperties.java). Which perhaps is just as well since that class doesn't appear to be very thread safe and that its giveMeRandomChannel method doesn't return a random channel as advertised but actually just returns the first open channel in the list of channels.

### Expected Behavior

I expect Eiffel REMReM Publish to perform well under reasonable loads and that it its performance will be contrained by CPU, memory, thread pool size, and other tunable parameters.

### The version of this project/repo, if applicable

master

### The version/edition of the Eiffel Protocol used, if applicable

N/A

**Comment by z-sztrom** [comment]
Confirmed, the synchronization sequence limits ```publish``` service to publish one event in a time, i.e. an event is published only when publishing of a previous one has been finished.

**Comment by z-sztrom** [comment]
I tried to compare current implementation of ``publish`` service with one without the ``synchronized`` section and I didn't observe any performance difference. Publishing took roughly the same time: non-synchronized version 1% faster.

I tested in Docker environment and I suspect that MB was a bottleneck. I'll continue testing.

@magnusbaeck, can you perform testing if I provide you non-synchronized version?

**Comment by z-sztrom** [comment]
I did testing using real machines (not Docker containers) and I found that publishing without synchronized section is cca 15-20% faster (publishing up to cca 100 events).

**Comment by magnusbaeck** [comment]
Sorry, I was away last week and have had to spend some time this week on getting back on track. Yes, I can do some load testing. 

100 events/s sounds very low. How many concurrent client connections did you run?

**Comment by z-sztrom** [comment]
No, it's a misunderstanding. Version without synchronization is 15-20% faster when publishing up to 100 events. Then the spped advantage drops down gradually and when publishing cca 10000 events, performance of both, synchronized and non-synchronized versions, becomes the same.

Overall performance I achieved on my setup was cca 3000events/sec.

**Comment by z-sztrom** [comment]
As you stated in your comment (https://github.com/eiffel-community/eiffel-remrem-publish/pull/219#discussion_r743612694), the problem arises from inside of semantics library. The implementation has never been prepared to work in parallel/concurrent environment.

Unfortunately we have MessageService interface defined and changing it would cause backward incompatibility. The only solution I have on my mind is to introduce thread-local variables to ensure that getHttpStatus() returns correct value.

Or should we alter the interface? What will be impacted?

**Comment by z-sztrom** [comment]
@magnusbaeck, fix of this issue is included in 2.1.6. Please, check. Thank you.
